sdl_viewer/src/terrain_drawer/layer.rs

```diff
@@ -0,0 +1,232 @@
+        );
+        assert!(
+            cur_pos.y < F64_MAX_INT && cur_pos.y > -F64_MAX_INT,
+            "Terrain location not representable."
+        );
+        self.u_terrain_pos.value = cur_pos.cast().unwrap();
+
+        let hori_strip = if moved.y > 0 {
+            self.load(
+                cur_pos.x,
+                self.terrain_pos.y + self.texture_size,
+                self.texture_size.try_into().unwrap(),
+                moved.y.try_into().unwrap(),
+            )
+        } else {
+            self.load(
+                cur_pos.x,
+                cur_pos.y,
+                self.texture_size.try_into().unwrap(),
+                moved.y.abs().try_into().unwrap(),
+            )
+        };
+        let vert_strip = if moved.x > 0 {
+            self.load(
+                self.terrain_pos.x + self.texture_size,
+                cur_pos.y,
+                moved.x.try_into().unwrap(),
+                self.texture_size.try_into().unwrap(),
+            )
+        } else {
+            self.load(
+                cur_pos.x,
+                cur_pos.y,
+                moved.x.abs().try_into().unwrap(),
+                self.texture_size.try_into().unwrap(),
+            )
+        };
+
+        self.heightmap.incremental_update(
+            moved.x as i32,
+            moved.y as i32,
+            vert_strip.height,
+            hori_strip.height,
+        );
+        self.colormap.incremental_update(
+            moved.x as i32,
+            moved.y as i32,
+            vert_strip.color,
+            hori_strip.color,
+        );
+    }
+
+    pub fn terrain_from_world(&self) -> Isometry3<f64> {
+        self.grid_coordinates.terrain_from_world.clone()
+    }
+
+    pub fn submit(&self) {
+        self.grid_coordinates.submit();
+        self.u_terrain_pos.submit();
+        self.heightmap.submit();
+        self.colormap.submit();
+    }
+
+    fn load(&self, min_x: i64, min_y: i64, width: usize, height: usize) -> HeightAndColor {
+        HeightAndColor {
+            height: self.height_tiles.load(min_x, min_y, width, height),
+            color: self.color_tiles.load(min_x, min_y, width, height),
+        }
+    }
+}
+
+/// This struct's job is to convert from continuous world positions to discrete
+/// terrain positions and to provide the necessary variables to do the inverse
+/// calculation to the vertex shader.
+struct GridCoordinateFrame {
+    u_origin: GlUniform<Vector3<f64>>,
+    u_world_from_terrain: GlUniform<Matrix4<f64>>,
+    u_resolution_m: GlUniform<f64>,
+    texture_half_extent: Vector2<i64>,
+    terrain_from_world: Isometry3<f64>,
+}
+
+impl GridCoordinateFrame {
+    fn new(program: &GlProgram, metadata: Metadata, texture_size: u32) -> Self {
+        let u_origin = GlUniform::new(&program, "terrain_origin_m", metadata.origin);
+        let u_world_from_terrain = GlUniform::new(
+            &program,
+            "terrain_to_world",
+            Matrix4::from({
```

Comment:
nit: I think if you compute `terrain_from_world` befpre `u_world_from_terrain` then you wouldn't have to clone `metadata.world_from_terrain` (and save 8 characters :P).

Reply:
Nice!